beamer/agent/chain.py

```diff
@@ -476,6 +489,117 @@ def maybe_invalidate(claim: Claim, context: Context) -> None:
     claim.start_challenge()
 
 
+def _proof_ready_for_l1_relay(request: Request) -> bool:
+    return (
+        not request.l1_resolved.is_active
+        and request.fill_tx is not None
+        and request.fill_timestamp is not None
+    )
+
+
+def _invalidation_ready_for_l1_relay(claim: Claim) -> bool:
+    return (
+        not claim.invalidated_l1_resolved.is_active
+        and claim.invalidation_tx is not None
+        and claim.invalidation_timestamp is not None
+    )
+
+
+def _l1_resolution_threshold_reached(claim: Claim, context: Context) -> bool:
+    l1_gas_cost = 1_000_000  # TODO: Adapt to real price
+    l1_gas_price = context.web3_l1.eth.gas_price
+    l1_safety_factor = 1.25
+    limit = int(l1_gas_cost * l1_gas_price * l1_safety_factor)
+
+    # Agent is claimer
+    if claim.claimer == context.address:
+        reward = int(claim.latest_claim_made.challenger_stake_total)
+    else:
+        # Agent is challenger
+        reward = claim.get_challenger_stake(context.address)
+        last_challenger = claim.latest_claim_made.last_challenger
+        if last_challenger == context.address:
+            reward -= (
+                claim.latest_claim_made.challenger_stake_total
+                - claim.latest_claim_made.claimer_stake
+            )
+    return reward > limit
+
+
+def _timestamp_is_l1_finalized(
+    timestamp: Timestamp, context: Context, target_chain_id: ChainId
+) -> bool:
+    # The internal memory should not be none because at this point
+    # the ChainUpdated event must have arrived, otherwise there would be no request
+    target_chain_finality = context.finality_periods[target_chain_id]
+    return int(time.time()) > timestamp + target_chain_finality
+
+
+@dataclass
+class _ResolveInfo:
```

Comment:
If we unify then we don't need this private class or the MESSAGES anymore. You can just inline in 552 to 563 and use local variables instead